File: 2020/src/day03.cpp
#include "day03.h"
#include <initializer_list>

using namespace common;

ResultBase* Day03::SolveFirst() const
{
	Grid<int> map = ParseInput();
	int res = CountTrees(map, Vec2(0, 0), Vec2(3, 1));
	return new Result(res);
}

ResultBase* Day03::SolveSecond() const
{
	Grid<int> map = ParseInput();
	int res = 1;
	for(Vec2 step : { Vec2(1, 1), Vec2(3, 1), Vec2(5, 1), Vec2(7, 1), Vec2(1, 2) })
	{
		res *= CountTrees(map, Vec2(0, 0), step);
	}
	return new Result(res);
}

int Day03::CountTrees(const Grid<int>& map, Vec2<int> start, Vec2<int> step) const
{
	int count = 0;
	while(start.y < map.Height())
	{
		if(map(start.x, start.y) != 0)
		{
			count++;
		}
		start += step;
		start.x %= map.Width();
	}
	return count;
}

Grid<int> Day03::ParseInput() const
{
	Grid<int> grid = Grid<int>(input.lines[0].length(), input.lines.size());
	for(size_t i = 0; i < input.lines.size(); i++)
	{
		for(size_t k = 0; k < input.lines[i].length(); k++)
		{
			grid.At(k, i) = input.lines[i][k] == '#';
		}
	}
	return grid;
}

File: 2025/src/day03.fs
module Day03
open Input
open Common

[<Literal>]
let name = "Lobby"

let parseInput (input: Input) = 
    input.lines
    |> List.map Seq.toList
    |> List.map (List.map (fun n -> uint64 (n - '0')))

let find (size: int) (digits: uint64 list) = 
    let rec findMaxDigit digits count = 
        match count with
        | 0 -> 0UL
        | _ -> 
            let limitIdx = List.length digits - count
            let max = digits.[..limitIdx] |> List.max
            let leftIdx = digits |> List.findIndex ((=) max)
            let shift = pown 10UL (count - 1)
            max * shift + (findMaxDigit (digits.[(leftIdx + 1) ..]) (count - 1))

    findMaxDigit digits size

let solve (selector: uint64 list -> uint64) (input: Input) = 
    input
    |> parseInput
    |> List.map selector
    |> List.sum

let solveFirst (input: Input) = input |> solve (find 2)
let solveSecond (input: Input) = input |> solve (find 12)
